fdbserver/workloads/Mako.actor.cpp

```diff
@@ -0,0 +1,530 @@
+			
+			std::string ts = format("T=%04.0fs: ", elapsed);
+			self->periodicMetrics.push_back(PerfMetric(ts + "Transactions/sec", (self->xacts.getValue() - last_xacts) / self->periodicLoggingInterval, false));
+			self->periodicMetrics.push_back(PerfMetric(ts + "Operations/sec", (self->totalOps.getValue() - last_ops) / self->periodicLoggingInterval, false));
+
+			last_xacts = self->xacts.getValue();
+			last_ops = self->totalOps.getValue();
+		}
+	}
+	ACTOR Future<Void> _setup(Database cx, MakoWorkload* self) {
+
+		state Promise<double> loadTime;
+		state Promise<vector<pair<uint64_t, double>>> ratesAtKeyCounts;
+
+		wait(bulkSetup(cx, self, self->rowCount, loadTime, self->insertionCountsToMeasure.empty(), self->warmingDelay,
+		               self->maxInsertRate, self->insertionCountsToMeasure, ratesAtKeyCounts));
+
+		// This is the setup time 
+		self->loadTime = loadTime.getFuture().get();
+		// This is the rates of importing keys
+		self->ratesAtKeyCounts = ratesAtKeyCounts.getFuture().get();
+
+		return Void();
+	}
+
+	ACTOR Future<Void> _start(Database cx, MakoWorkload* self) {
+		vector<Future<Void>> clients;
+		for (int c = 0; c < self->actorCountPerClient; ++c) {
+			clients.push_back(self->makoClient(cx, self, self->actorCountPerClient / self->transactionsPerSecond, c));
+		}
+
+		if (self->enableLogging)
+			clients.push_back(tracePeriodically(self));
+
+		wait( timeout( waitForAll( clients ), self->testDuration, Void() ) );
+		return Void();
+	}
+
+	ACTOR Future<Void> makoClient(Database cx, MakoWorkload* self, double delay, int actorIndex) {
+
+		state Key rkey;
+		state Value rval;
+		state Transaction tr(cx);
+		state bool doCommit;
+		state int i, count;
+		state uint64_t range, indBegin, indEnd, rangeLen;
+		state double lastTime = now();
+		state double commitStart;
+		state KeyRangeRef rkeyRangeRef;
+		state vector<int> perOpCount(MAX_OP, 0);
+
+		TraceEvent("ClientStarting").detail("AcotrIndex", actorIndex).detail("ClientIndex", self->clientId).detail("NumActors", self->actorCountPerClient);
+
+		loop {
+			// used for throttling
+			wait(poisson(&lastTime, delay));
+			try{
+				// user-defined value: whether commit read-only ops or not; default is false
+				doCommit = self->commitGet;
+				for (i = 0; i < MAX_OP; ++i) {
+					if (i == OP_COMMIT) 
+						continue;
+					for (count = 0; count < self->operations[i][0]; ++count) {
+						// generate random key-val pair for operation
+						rkey = self->ind2key(self->getRandomKey(self->rowCount));
+						rval = self->randomValue();
+
+						range = (self->operations[i][1] > 0) ? self->operations[i][1] : 1;
+						rangeLen = digits(range);
+						indBegin = self->getRandomKey(self->rowCount);
+						// KeyRangeRef(min, maxPlusOne)
+						indEnd = std::min(indBegin + range, self->rowCount);
+						rkeyRangeRef = KeyRangeRef(self->ind2key(indBegin), self->ind2key(indEnd));
+
+						if (i == OP_GETREADVERSION){
+							state double getReadVersionStart = now();
```

Comment:
Is there a reason, `getReadVersion` can't be wrapped into a `logLatency` actor? This looks a bit inconsistent to me...

Reply:
Will fix it.